File: src/main/java/org/willxu/algorithm/service/bool/ContainsDuplicate.java
/*
 * Copyright (C) 2021  Weiyang(Will) Xu
 * 
 * This file is part of algorithm.
 * Algorithm is free software: you can redistribute it and/or modify
 * it under the term of the GNU General Public License version 3 or
 * any later version, as specified in the readme.md file.
 */

package org.willxu.algorithm.service.bool;

public interface ContainsDuplicate {
    /**
     * Given an integer array nums, return true if any value appears at
     * least twice in the array, and return false if every element is
     * distinct.
     *
     * @param nums 1 <= nums.length <= 10^5
     *             -10^9 <= nums[i] <= 10^9
     * @return contains duplicate or not
     */
    boolean containsDuplicate(int[] nums);
}

File: src/leetcode/p0200_0299/p0217_contains_duplicate/contains_duplicate.py
from abc import ABC, abstractmethod
from typing import List

class ContainsDuplicate(ABC):
    """
    Given an integer array nums, return true if any value appears at least twice in the array, and return false if every element is distinct.
    """
    @abstractmethod
    def containsDuplicate(self, nums: List[int]) -> bool:
        """
        Args:
            nums:
                - 1 <= nums.length <= 10^5
                - -10^9 <= nums[i] <= 10^9
        """
        pass

class ContainsDuplicateSet(ContainsDuplicate):
    def containsDuplicate(self, nums: List[int]) -> bool:
        seen = set()
        for num in nums:
            if num in seen:
                return True
            seen.add(num)
        return False
